File: src/main/java/net/projecteuler/barreiro/problem/Solver036.java
/* COPYRIGHT (C) 2014 barreiro. All Rights Reserved. */

package net.projecteuler.barreiro.problem;

import static net.projecteuler.barreiro.algorithm.Combinatorics.palindromeStream;
import static net.projecteuler.barreiro.algorithm.util.LongUtils.isPalindrome;

/**
 * The decimal number, 585 = 10010010012 (binary), is palindromic in both bases.
 * <p>
 * Find the sum of all numbers, less than one million, which are palindromic in base 10 and base 2.
 * <p>
 * (Please note that the palindromic number, in either base, may not include leading zeros.)
 *
 * @author barreiro
 */
public class Solver036 extends ProjectEulerSolver {

    public Solver036() {
        this( 1000000 );
    }

    public Solver036(long n) {
        super( n );
    }

    // --- //

    public long solve() {
        return palindromeStream( N ).filter( p -> isPalindrome( p, 2 ) ).sum();
    }

}

File: src/main/rust/euler/solver036.rs
// COPYRIGHT (C) 2017 barreiro. All Rights Reserved.
// Rust solvers for Project Euler problems

use algorithm::cast::to_i64;
use algorithm::digits::{Digits, palindromes};
use algorithm::filter::less_than_u64;
use Solver;

/// The decimal number, `585 = 1001001001 (binary)`, is palindromic in both bases.
///
/// Find the sum of all numbers, less than one million, which are palindromic in base `10` and base `2`.
///
/// (Please note that the palindromic number, in either base, may not include leading zeros.)
pub struct Solver036 {
    pub n: u64,
}

impl Default for Solver036 {
    fn default() -> Self {
        Self { n: 1_000_000 }
    }
}

impl Solver for Solver036 {
    fn problem_name(&self) -> &str { "Double-base palindromes" }

    fn solve(&self) -> i64 {
        palindromes().take_while(less_than_u64(self.n)).filter(|&p| Digits::from((p, 2)).is_palindrome()).map(to_i64).sum()
    }
}
